### ISSUE TITLE
(Almost) arbitrary functions are assignable to an assertion-function-typed variable

### ISSUE THREAD

**Issue by jiftechnify** [microsoft/TypeScript#49262]
# Bug Report

### 🔎 Search Terms

assertion functions assignability inconsistent type guards

### 🕗 Version & Regression Information

- This is the behavior in every version I tried, and I reviewed the FAQ for entries about type system behavior

### ⏯ Playground Link

[Playground link with relevant code](https://www.typescriptlang.org/play?ts=4.6.4#code/MYewdgzgLgBAhhCBTATlAcgVwLYCNUwC8MAFAB4BcMmYA1mCAO5gCUVCyaEMZMAltzA58KIgD4YAbwBQMfgDNSUAJ4AHJCEW8AhIWIByIXlT6WU2XJhQAFiiYwAoijsoSAIgaw4MIyLcsLAF9pYIB6UJgITAhVPmA+EGj4RD4AczBsJDAoADo86VBIL0RUKABlKBQ+MFSqcioaeiZWcWTOKG5eAUjK6tSiNtKsYxRpAvBoHga6BmYBgEYABkWAbmkOUoqqmvIWFZhwmAAhJGA4aKQYAAMNtC2+q-5uOGAoTDgAGw-la9uMYVQVwAND4QDBUC4nlZbM1pKEAFSInj6bgqdRQ2Y9bb9RHw0LSMg5KK4aDYkjzPYHCIMGCgbCxD6XCEgFAg3CYWAoGhQPiZcHOFnaMbjIoweQgMHEepYvowAA+1DAABMkPJqkglWwfADRIQJEtFnCIlEYnEEkkOGkMllcvlCpM-gAVLk2ZR1SgymryxUqtVgDVmPWDLg8AbikBje2wN2e-oKmi+9VKsOfZAwAD8MH0cH0MGmif9SrWTpd1mUJGUe2kyiJmBJvR2FP2hxpdIZTIFrJg7M53N5HZcQrGCJg6BAUCQVHw1jgADcEihuJorGpLql3igldw8UaYI7VzBl1cBABVGbNR7YOBQYDWJDcNyq+SnHmzpBuFfoo+n8-MABMwIwLuPbwMAwAcu8Xw-MeEBnk0zCPN0nhtFacC4IyVhgjBcGzGAAHdqc5xpjYlwQKhbwoJc36wb+YCIYI47wJ+lyqJRSpxNeSA5CKkw-vBYADNKjS4VquASoycACUGMhyJRFECZUmBICEPGwHxuF-u60z8YGEhdNwwnNAM6mwru+7ouucCbs8lG0iAs5WXwkmwIwfA2DAlExC+mEwG4nmYB8sBou+pB8DkXHXKSDz8AJJB5Dkumht0UU1Cw3FRk8ACCLJHIJHqGcwWr6VmObevouC5tJFhyZgKACbwegGKVcoKg1ehZhVKmqU89w1Fpio6a0xUpf0xACNlKBHN1439QVLRBsV2aVVlOXSEAA)

### 💻 Code

```ts
const assertNumber = (x: unknown): asserts x is number => {
  if (typeof x !== 'number') {
    throw Error("not a number")
  }
}
// suspicious assignment...
const assertString: (x: unknown) => asserts x is string = assertNumber

const x: unknown = 100;
assertString(x); // Because `assertString` is actually `assertNumber`, no error is thrown
/*** x's type is now string ***/
x.substring(1); // no compile error, but runtime error!


const foo = (x: string | undefined): number => 100
// suspicious assignment...
const assertTruthy: (x: string | undefined) => asserts x = foo

const y: string | undefined = false ? 'a' : undefined;
assertTruthy(y);
y.substring(1); // no compile error, but runtime error!
```

### 🙁 Actual behavior

Both (suspicious) assignments succeed with no compile error.  This behavior is not consistent with that of type guards.

### 🙂 Expected behavior

Both assignments should be rejected by the compiler.

I believe assignability rule for assertion functions should be consistent with the rule for type guards.  At least, an assignment of a function which doesn't have assertion function signature to an assertion-function-typed variable should be rejected.  Furthermore, for assertion function signatures with "result type" (`(x: ...) => asserts x is T`), it should be covariant with respect to the "result type".

**Comment by RyanCavanaugh** [comment]
> At least, an assignment of a function which doesn't have assertion function signature to an assertion-function-typed variable should be rejected.

Original notes at #32695 don't mention this, but IIRC the motivating behavior for this logic was that we wanted a binding with an assertion function type to be intializable by a "normal" function, i.e.:
```ts
const assertNumber: (x: unknown) => asserts x is number = x => {
  if (typeof x !== 'number') {
    throw Error("not a number")
  }
}
```
here the return type of the function expression is just `void`, so the initialization is allegedly illegal under this rule unless we added new logic to e.g. contextually-modify the return type of the function expression.

Assigning something with two different asserted-types is super sketchy. We could potentially flag this, but I think the lived experience over the last three years is that aliasing assertion functions of any kind is extraordinarily rare since they don't have any meaningful substitutability or polymorphism.

**Comment by bgourlie** [comment]
>aliasing assertion functions... don't have any meaningful substitutability or polymorphism.

I disagree and believe this is a legitimate soundness issue! Consider the following, which among few other things defines a generic array assertion function:

```
type AssertFn<T> = (val: unknown) => asserts val is T;

const str: AssertFn<string> = (val) => {
  if (typeof val !== 'string') {
    throw new Error('str assertion failed');
  }
}

const num: AssertFn<number> = (val) => {
  if (!Number.isFinite(val)) {
    throw new Error('num assertion failed');
  }
}

const array: <T>(assert: AssertFn<T>) => AssertFn<T[]> =
  (assert) => (val) => {
    if (!Array.isArray(val)) {
      throw new Error('assertIsArray failed');
    }
    val.forEach(assert);
  };

const assertStringArray: AssertFn<string[]> = array(num);
```
[Link to playground](https://www.typescriptlang.org/play?ssl=23&ssc=56&pln=1&pc=1#code/C4TwDgpgBAggznCAnYAxAdgHgCoD4oC8UAFAG4CGANgFxQCu6A1ugPYDu6AlIfuQssDhQKlKAEsh2ANwAoGQGMW6OMCgqkteIhQZM6segDm+ImSrcC+AN4yo4gGYlQkFo5FQAhASIByfUZ9uGzs7YAALJHYodAg2KABRJEikYj9gJCg+bWAxJSh7cjFKCAATQNk7AF8ZaoUlFWi6AFtNfh0sdGaAI2QTEhELa1sHEg8AOW7kADoJVAMxYAgzSk4g4dCIqJi4xOTUzqbMtpy8gqLS8uHq2sVlVXIk8hBaHFxiLIFW7N08Qdhjn4AbQAuiZhu9jn9ln9giExI5iB4YI8QDM4MikE9oWsQiFwpE4tsEkkWCkfB8UABJdEo-KFYplTgVELVXEiKb2UnxcjyMIQ7JMq6yOp3I7ZADK6QMhgxTy+Al0-mMhEyKOIB0Fci1MiAA)

Note that I explicitly type `assertStringArray` as `AssertFn<string[]>` but am actually supplying  a `AssertFn<number[]>` via `array(num)`, and the compiler is fine with this. If I remove the explicit type annotation, it correctly infers the type to `AssertFn<number>`.

**Comment by acutmore** [comment]
One potential reason why this could be particularly 'dangerous' is that it's required to specify the assertion types to ensure they impact CFA.

```ts
type TypePredicate<T> = (v: unknown) => v is T;
type TypeAssertion<T> = (v: unknown) => assertions v is T;

declare function makeAssertion<T>(predicate: TypePredicate<T>): TypeAssertion<T>;
declare function isString(v: unknown): v is string;
```

```ts
const assertIsString = makeAssertion(isString);
declare let x: unknown;
assertIsString(x);
x; // still unknown - need to add type annotation above...
```

```diff
- const assertIsString = makeAssertion(isString);
+ const assertIsString: TypeAssertion<string> = makeAssertion(isString);
declare let x: unknown;
assertIsString(x);
x; // good, x is string
```

but if we put the wrong type...

```diff
- const assertIsString = makeAssertion(isString);
+ const assertIsString: TypeAssertion<number> = makeAssertion(isString);
declare let x: unknown;
assertIsString(x);
x; // bad, x is number
```

I was expecting `const assertIsString: TypeAssertion<number> = makeAssertion(isString);` to produce a type error.

**Comment by mwoenker** [comment]
This behavior was very surprising to me. I've been working on a project that validated JSON passed to API endpoints using assertion functions. I was shocked to realize that the code actually had no type safety. No one else on the project realized it either, including those who wrote the original code. A very simplified version of the code in question looks like this:

```
type Validator<T> = (val: unknown) => asserts val is T
type Handler<T> = (request: T) => void;

function attachApiEndpoint<T>(path: string, handler: Handler<T>, assertValid: Validator<T>) {
  server.attach(path, (request) => { 
    assertValid(request.requestBody);
    handler(request.requestBody);
  })
}

const assertString: Validator<string> = (val: unknown): asserts val is string => {
  if (typeof val !== 'string') {
    throw new Error('not a string');
  };
}

attachApiEndpoint(
  '/path/to/endpoint',
  (body: number) => {
    // body will actually be a string here!
  },
  assertString // no type error prevents an assertion not matching the handler from being passed here
)
```
